webapp/src/js/services/db-sync.js

```diff
@@ -46,14 +47,67 @@ angular
       );
     };
 
+    // Retry replication for every "real" rev X times
+    // we enable retrying by "touching" the doc, pushing it to the end of the changes feed
+    // we store the rev of the doc we touch and we only increase the replication_retry if the calculated previous
+    // rev matches the previous retry rev. This ensures that external updates (for example user updates) would reset
+    // the retry counter.
+    const retryForbiddenFailure = err => {
+      if (!err || !err.id) {
+        return;
+      }
+
+      const getPreviousRev = doc => {
+        const revisions = doc._revisions;
+        if (!revisions) {
+          return;
+        }
+
+        if (revisions.start <= 1 || revisions.ids.length <= 1) {
+          return;
+        }
+
+        return `${revisions.start - 1}-${revisions.ids[1]}`;
+      };
+
+      return DB()
+        .get(err.id, { revs: true })
+        .then(doc => {
+          doc.replication_retry = doc.replication_retry || {};
+          doc.replication_retry.count = doc.replication_retry.count || 1;
+
+          if (doc.replication_retry.rev) {
+            const previousRev = getPreviousRev(doc);
+            const consecutiveAttempts = previousRev && previousRev === doc.replication_retry.rev;
+            if (!consecutiveAttempts) {
+              doc.replication_retry.count = 1;
+            } else {
+              doc.replication_retry.count += 1;
+            }
+          }
+
+          if (doc.replication_retry.count > MAX_REPLICATION_RETRY_COUNT) {
+            return;
+          }
+
+          doc.replication_retry.rev = doc._rev;
+          delete doc._revisions;
+          return DB().put(doc);
+        })
+        .catch(err => {
+          $log.error(`Error when retrying replication for forbidden doc`, err);
+        });
+    };
```

Comment:
I think this is complicated enough that it should be its own service. This would make your unit testing easier because you just need to verify that the service is called from DbSync when the right error is thrown, and then your new unit can focus on just what happens when this is called.